File: frontend/lib/clients.ts
import { API_V1_URL } from "./auth";
import { request } from "./api";
import { createCrudResource } from "./crud";
import type { Patient } from "./patients";

export type NotificationChannel = "email" | "sms";

export type Client = {
  id: number;
  first_name: string;
  last_name: string;
  email: string;
  phone_number: string;
  street: string;
  city: string;
  postal_code: string;
  notes: string;
  preferred_notification_channel: NotificationChannel;
  created_at: string;
  updated_at: string;
};

export type ClientInput = {
  first_name: string;
  last_name: string;
  email: string;
  phone_number: string;
  street: string;
  city: string;
  postal_code: string;
  notes?: string;
  preferred_notification_channel?: NotificationChannel;
};

const CLIENTS_URL = `${API_V1_URL}/clients/`;

const clients = createCrudResource<Client, ClientInput>(CLIENTS_URL);

export const listClients = clients.list;
export const listAllClients = clients.listAll;
export const getClient = clients.get;
export const createClient = clients.create;
export const updateClient = clients.update;
export const deleteClient = clients.remove;

export type ClientWithPatients = Client & { patients: Patient[] };

export function getClientFull(id: number) {
  return request<ClientWithPatients>(`${CLIENTS_URL}${id}/full/`);
}

File: backend/src/repositories/clients.py
from clients.models import Client as ClientModel

from src.models.clients import Client, NotificationChannel
from src.repositories.base import DjangoRepository


def client_to_dataclass(obj: ClientModel) -> Client:
    return Client(
        id=obj.id,
        first_name=obj.first_name,
        last_name=obj.last_name,
        email=obj.email,
        phone_number=obj.phone_number,
        street=obj.street,
        city=obj.city,
        postal_code=obj.postal_code,
        notes=obj.notes,
        preferred_notification_channel=NotificationChannel(
            obj.preferred_notification_channel
        ),
        created_at=obj.created_at,
        updated_at=obj.updated_at,
    )


class ClientRepository(DjangoRepository[Client]):
    model = ClientModel
    to_dataclass = staticmethod(client_to_dataclass)
